sys/pm_layered/pm.c

```diff
@@ -36,52 +36,52 @@
 /**
  * @brief Global variable for keeping track of blocked modes
  */
-volatile pm_blocker_t pm_blocker = { .val_u32 = PM_BLOCKER_INITIAL };
+static pm_blocker_t pm_blocker = { .val_u32 = PM_BLOCKER_INITIAL };
 
 void pm_set_lowest(void)
 {
-    pm_blocker_t blocker = pm_blocker;
+    unsigned state = irq_disable();
     unsigned mode = PM_NUM_MODES;
     while (mode) {
-        if (blocker.val_u8[mode-1]) {
+        if (pm_blocker.val_u8[mode-1]) {
             break;
         }
         mode--;
     }
 
-    /* set lowest mode if blocker is still the same */
-    unsigned state = irq_disable();
-    if (blocker.val_u32 == pm_blocker.val_u32) {
-        DEBUG("pm: setting mode %u\n", mode);
-        pm_set(mode);
-    }
-    else {
-        DEBUG("pm: mode block changed\n");
-    }
+    pm_set(mode);
     irq_restore(state);
 }
 
 void pm_block(unsigned mode)
 {
-    assert(pm_blocker.val_u8[mode] != 255);
-
     unsigned state = irq_disable();
+    assert(pm_blocker.val_u8[mode] != 255);
     pm_blocker.val_u8[mode]++;
     irq_restore(state);
 }
 
 void pm_unblock(unsigned mode)
 {
+    unsigned state = irq_disable();
     assert(pm_blocker.val_u8[mode] > 0);
+    pm_blocker.val_u8[mode]--;
+    irq_restore(state);
+}
 
+pm_blocker_t pm_get_blocker(void)
+{
+    pm_blocker_t result;
     unsigned state = irq_disable();
-    pm_blocker.val_u8[mode]--;
+    result = pm_blocker;
     irq_restore(state);
+    return result;
```

Comment:
This seems overly complicated, you want to create a copy for it, right?

Reply:
The read to the stack variable `result` is guarded in `irq_disable()` and `irq_restore()`. This prevents e.g. on AVR devices that the read sequence is interrupted. Direct access to the variable could result in reading an inconsistent state, as some bytes could be retrieved at time A and the remaining bytes at time B. The result returned would be a blocker state that is potentially different to both time A and time B. This is something we have to prevent.

Additionally: `irq_disable()` and `irq_restore()` are memory barriers, which prevent elimination of seemingly unessecary reads.

E.g. a thread calling `pm_get_blocker()` in a loop and printing the result every 1 second. Without the memory barriers, the compiler would assume that the value read in the first iteration of the loop is still the same as in the second, as it is not changed by the current thread of execution. But as other threads or ISRs could have changed it, it is needed to be read again. https://lwn.net/ recently had a nice serious about compiler optimizations that are fully valid in terms of the C standard but render your program unusable, unless the programmer uses techniques like memory barriers to tell the compiler which time of optimizations are not applicable in a specific context.